### DISCUSSION TITLE
Can you tell me the how we can encrypt blade file?

### DISCUSSION THREAD

**Discussion by vishalmicrolevel** [laravel/framework#44916] [General]
Can you tell me the how we can encrypt blade file?

Without encrypt blade file everyone know the my code. so give me proper solution.

**Comment by Sophist-UK** [comment]
@vishalmicrolevel I am entirely unclear why you would want to encrypt your blade file.

If you want to offer a Laravel-based app on commercial terms for users to run on their own servers, by definition you have to provide the source code to them. As far as I can see, the options you have are:

  1. Obfuscate your code. You run your code through a tool that replaces all the names which mean anything with new names which are either very short, (e.g. A, B, ... Z, AA, AB, ... ZZ, AAA ...) or random characters (IKTRHEOSNEPT). By doing this you make it **very** difficult for anyone to understand what your code is doing. This would probably need to be a Laravel-specific obfuscator tool - I have not researched this, so I have no idea whether one exists. But beware, that if your user experiences a bug, the exception / log / traceback messages will use these funny names, so tracking down where the bug is may be a lot harder.
  2. Rewrite your application in a compiled or semi-compiled language (like C or Python) where you can ship machine or byte code to the user rather than source code. Machine / byte code is a much lower-level way of obfuscating your code. (As an aside, modern PHP versions do optimisation which creates bytecode or machine code for some / all of your application - but AFAIK, there is no way of extracting and shipping this in place of the source code.)
  3. Offer your app only on your own servers as a service. You do not need to expose the source code at all in this case.

Or have I misunderstood the root of your issue?

**Reply by vishalmicrolevel** [reply]
I want to set up code for my client's desktop. but he is a little bit knowledgeable about Laravel.
So I don't want to give the original code to him. otherwise, he will sell my code also.
So I want the encrypt full Laravel code.
Using this package https://github.com/SiavashBamshadnia/Laravel-Source-Encrypter <= Encrypt PHP code.
but I m not able to encrypt the blade file code.
So using the blade file anyone knows this variable whose value is stored? so I don't want them.
So please give me a proper solution

**Reply by Sophist-UK** [reply]
The issue is that the way that this package decrypts the PHP code is by using the phpBolt PHP plugin which decrypts the executable instructions inside php. This is what makes it secure - the encryption key and the encryption/decryption code can only be accessed by the PHP internals, and AFAIK there is no way to request PHP to decrypt a file so that it can be read as text (because if there were, the files would be trivially decryptable by writing a PHP script to do so.

The Blade files are (I think - I haven't checked the code) read by Laravel's Blade code and turned into a (cacheable) native php file. That is why it cannot be encrypted.

If you look inside the [Laravel-Source-Encrypter code](https://github.com/SiavashBamshadnia/Laravel-Source-Encrypter/blob/master/src/SourceEncryptCommand.php#L120-L128) you can see that it explicitly excludes `*.blade.php` files from being encrypted.

If you change your code to check for a pre-compiled blade file in  a subdirectory and use that rather than blade if it exists, it should (in theory at least) be possible to ship your code only with a compiled version of the blade files and not the `*.blade.php` files themselves.

**Reply by Sophist-UK** [reply]
I have raised a feature request for this at https://github.com/SiavashBamshadnia/Laravel-Source-Encrypter/issues/28.

**Reply by Sophist-UK** [reply]
P.S. If you genuinely think your client will steal your code and resell it, you really need either to have a stricter license, with bigger penalties for reselling the code on, or to refuse to sell your code to them in the first place.
